pkg/ratelimit/limitware/middleware.go

```diff
@@ -0,0 +1,216 @@
+// Package limitware provides middleware for rate limiting HTTP handlers.
+package limitware
+
+import (
+	"context"
+	"crypto/sha1"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+
+	"github.com/google/exposure-notifications-server/pkg/logging"
+
+	"github.com/sethvargo/go-limiter"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/api/metric"
+)
+
+const (
+	// HeaderRateLimitLimit, HeaderRateLimitRemaining, and HeaderRateLimitReset
+	// are the recommended return header values from IETF on rate limiting. Reset
+	// is in UTC time.
+	HeaderRateLimitLimit     = "X-RateLimit-Limit"
+	HeaderRateLimitRemaining = "X-RateLimit-Remaining"
+	HeaderRateLimitReset     = "X-RateLimit-Reset"
+
+	// HeaderRetryAfter is the header used to indicate when a client should retry
+	// requests (when the rate limit expires), in UTC time.
+	HeaderRetryAfter = "Retry-After"
+)
+
+// KeyFunc is a function that accepts an http request and returns a string key
+// that uniquely identifies this request for the purpose of rate limiting.
+//
+// KeyFuncs are called on each request, so be mindful of performance and
+// implement caching where possible. If a KeyFunc returns an error, the HTTP
+// handler will return Internal Server Error and will NOT take from the limiter
+// store.
+type KeyFunc func(r *http.Request) (string, error)
+
+// IPKeyFunc returns a function that keys data based on the incoming requests IP
+// address. By default this uses the RemoteAddr, but you can also specify a list
+// of headers which will be checked for an IP address first (e.g.
+// "X-Forwarded-For"). Headers are retrieved using Header.Get(), which means
+// they are case insensitive.
+func IPKeyFunc(headers ...string) KeyFunc {
+	return func(r *http.Request) (string, error) {
+		for _, h := range headers {
+			if v := r.Header.Get(h); v != "" {
+				return v, nil
+			}
+		}
+
+		ip, _, err := net.SplitHostPort(r.RemoteAddr)
+		if err != nil {
+			return "", err
+		}
+		return ip, nil
+	}
+}
+
+// Middleware is a handler/mux that can wrap other middlware to implement HTTP
+// rate limiting. It can rate limit based on an arbitrary KeyFunc, and supports
+// anything that implements limiter.Store.
+type Middleware struct {
+	store      limiter.Store
+	keyFunc    KeyFunc
+	meter      metric.Meter
+	reqCounter metric.Int64Counter
+}
+
+// NewMiddleware creates a new middleware suitable for use as an HTTP handler.
+// This function returns an error if either the Store or KeyFunc are nil.
+func NewMiddleware(s limiter.Store, f KeyFunc) (*Middleware, error) {
+	if s == nil {
+		return nil, fmt.Errorf("store cannot be nil")
+	}
+
+	if f == nil {
+		return nil, fmt.Errorf("key function cannot be nil")
+	}
+
+	meter := global.Meter("github.com/sethvargo/go-limiter")
```

Comment:
Is this the right meter? Or would we want to use this middleware package name instead?

Reply:
Changed, good call.